specs/core/1_new_shards.md

```diff
@@ -0,0 +1,546 @@
+    all_pubkeys = []
+    all_message_hashes = []
+    aggregation_bits = indexed_attestation.attestation.aggregation_bits
+    assert len(aggregation_bits) == len(indexed_attestation.committee)
+    for i, custody_bits in enumerate(indexed_attestation.attestation.custody_bits):
+        assert len(custody_bits) == len(indexed_attestation.committee)
+        for participant, abit, cbit in zip(indexed_attestation.committee, aggregation_bits, custody_bits):
+            if abit:
+                all_pubkeys.append(state.validators[participant].pubkey)
+                # Note: only 2N distinct message hashes
+                all_message_hashes.append(hash_tree_root(
+                    AttestationCustodyBitWrapper(hash_tree_root(indexed_attestation.data), i, cbit)
+                ))
+            else:
+                assert cbit == False
+        
+    return bls_verify_multiple(
+        pubkeys=all_pubkeys,
+        message_hashes=all_message_hashes,
+        signature=indexed_attestation.signature,
+        domain=get_domain(state, DOMAIN_BEACON_ATTESTER, indexed_attestation.data.target.epoch),
+    )
+```
+
+### `get_attestation_shard`
+
+```python
+def get_shard(state: BeaconState, attestation: Attestation) -> Shard:
+    return Shard((attestation.data.index + get_start_shard(state, data.slot)) % ACTIVE_SHARDS)
+```
+
+## Beacon Chain Changes
+
+### New beacon state fields
+
+```python
+    shard_states: Vector[ShardState, MAX_SHARDS]
+    online_countdown: Bytes[VALIDATOR_REGISTRY_LIMIT]
+    current_light_committee: CompactCommittee
+    next_light_committee: CompactCommittee
+```
+
+### New beacon block data fields
+
+```python
+    shard_transitions: Vector[ShardTransition, MAX_SHARDS]
+    light_client_signature_bitfield: Bitlist[LIGHT_CLIENT_COMMITTEE_SIZE]
+    light_client_signature: BLSSignature
+```
+
+### Attestation processing
+
+#### `validate_attestation`
+
+```python
+def validate_attestation(state: BeaconState, attestation: Attestation) -> None:
+    data = attestation.data
+    assert data.index < ACTIVE_SHARDS
+    shard = get_shard(state, attestation)
+    proposer_index = get_beacon_proposer_index(state)
+
+    # Signature check
+    assert is_valid_indexed_attestation(state, get_indexed_attestation(state, attestation))
+    # Type 1: on-time attestations
+    if data.custody_bits != []:
+        # Correct slot
+        assert data.slot == state.slot
+        # Correct data root count
+        start_slot = state.shard_next_slots[shard]
+        offset_slots = [start_slot + x for x in SHARD_BLOCK_OFFSETS if start_slot + x < state.slot]
+        assert len(attestation.custody_bits) == len(offset_slots)
+        # Correct parent block root
+        assert data.beacon_block_root == get_block_root_at_slot(state, state.slot - 1)
+    # Type 2: delayed attestations
+    else:
+        assert state.slot - slot_to_epoch(data.slot) < EPOCH_LENGTH
+        assert data.shard_transition_root == Hash()
```

Comment:
- Is it a kind of reduced form of attestation data?
- If so, does that mean the late attesters have to rebuild `AttestationData` of the old slot with empty `custody_bits` and `shard_transition_root`, and broadcast it?
- It seems the proposer can help transform ReducedAttestation in the [previous version](https://github.com/ethereum/eth2.0-specs/commit/0469d57ea26c848ffc2864741cc12d7dba2986c4)?

Reply:
Yes, it's a reduced form, and late attesters do have to rebroadcast.